### PR TITLE
e2e: chain initialization refactor for extensibility

### PR DESCRIPTION
**Pull request by p0mvn**
Part of: https://github.com/osmosis-labs/osmosis/issues/1879

## What is the purpose of the change

Our current chain initialization design does not allow for spinning up single nodes. This functionality is useful for testing feautes like state-sync.

This PR introduced the barebones structure to initialize single nodes via Docker. To build a single node initialization image, run:
```
make docker-build-e2e-init-node
```
Additional information about the new design can be found in `tests/e2e/initialization/README.md`

## Brief Changelog

- move initialization logic to `initialzation` package
- refactor init Dockerfile to support scripts provided by `ARG`
- create a stub script for single node initialization

## Testing and Verifying

This change is a trivial rework / code cleanup without any test coverage.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? 
      * https://github.com/osmosis-labs/osmosis/blob/0f3677649a43ca8ea6a906f49fb70d4565b514e9/tests/e2e/initialization/README.md

### CODE REVIEW

File: tests/e2e/initialization/node/main.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package main
+
+// Single-node initialization is not implemented yet This stub is created
+// as an intermediary step before final result.
+func main() {
+	panic("not implemented")
```
Comment (p0mvn) [inline tests/e2e/initialization/node/main.go:6]: This is a stub for now. The logic is to be added in a separate PR.

This is how we are going to spin up nodes for testing state-sync

File: tests/e2e/initialization/init.Dockerfile [hunk 2]
```diff
@@ -22,14 +24,22 @@ RUN sha256sum /lib/libwasmvm_muslc.x86_64.a | grep f6282df732a13dec836cda1f399dd
...
+COPY --from=build /osmosis/build/${E2E_SCRIPT_NAME} /bin/${E2E_SCRIPT_NAME}
 
 ENV HOME /osmosis
 WORKDIR $HOME
 
-ENTRYPOINT [ "chain_init" ]
+# Docker ARGs are not expanded in ENTRYPOINT in the exec mode. At the same time,
+# it is impossible to add CMD arguments when running a container in the shell mode.
+# As a workaround, we create the entrypoint.sh script to bypass these issues.
+RUN echo "#!/bin/bash\n${E2E_SCRIPT_NAME} \"\$@\"" >> entrypoint.sh && chmod +x entrypoint.sh
```
Comment (czarcas7ic) [inline tests/e2e/initialization/init.Dockerfile:43]: For my learning, what are the semantics here `\"\$@\""`? I have never seen this before.

File: tests/e2e/initialization/README.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+we might want to spin up a chain of v8 and test the upgrade.
+
+Additionally, there are known file permission errors when initializing
+configurations as non-root. This is troublesome both in CI and locally.
+Doing this via Docker instead, allows us to initialize these files as
+a root user, bypassing the file permission issues.
+
+## Structure
+
+Each folder in `tests/e2e/initialization` corresponds to a standolone script.
```
Comment (czarcas7ic) [inline tests/e2e/initialization/README.md:21]:
```suggestion
Each folder in `tests/e2e/initialization` corresponds to a standalone script.
```

**Review by czarcas7ic — APPROVED** [review]
Personally untested, but all of the logic here is sound to me. Well done 👍

File: tests/e2e/initialization/node/main.go (hunk 1 above)
Comment (ValarDragon) [reply tests/e2e/initialization/node/main.go:6]: Err, why a main package at all?

Want to go ahead and add this in whichever PR you add main package logic?

**Review by ValarDragon — APPROVED** [review]
LGTM aside from lack of udnerstanding the added main package

File: tests/e2e/initialization/node/main.go (hunk 1 above)
Comment (p0mvn) [reply tests/e2e/initialization/node/main.go:6]: Chain initialization and node initialization are run as scripts via Docker. `main` package is the only entry point to a Go program. That's why we need it

chain init is a more complete example of how this is going to look in the end:
https://github.com/osmosis-labs/osmosis/blob/main/tests/e2e/chain_init/main.go

And yes, I can remove and readd this in another PR

File: tests/e2e/containers/containers.go [hunk 4]
```diff
@@ -88,6 +91,16 @@ func (m *Manager) ExecCmd(t *testing.T, chainId string, validatorIndex int, comm
...
+			errBufString := errBuf.String()
+
+			// Note that this does not match all errors.
+			// This only works if CLI outpurs "Error" or "error"
+			// to stderr.
+			if errRegex.MatchString(errBufString) {
+				t.Log(errBufString)
+				return false
+			}
+
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply tests/e2e/containers/containers.go:103]: I thought that there are success results being returned via stderr, are they not? 

From my understanding, if we return for anything printed to stderr, we might not get these results

File: tests/e2e/initialization/node/main.go (hunk 1 above)
Comment (p0mvn) [reply tests/e2e/initialization/node/main.go]: Done

File: tests/e2e/initialization/init.Dockerfile (hunk 2 above)
Comment (p0mvn) [reply tests/e2e/initialization/init.Dockerfile:43]: This is bash syntax for expanding cmd line parameters.  

Basically, since we want to call chain init and node init scripts by providing them with certain initialization arguments via Docker, we need to be able to expand them in the bash script.

This post points to a relevant man page about the syntax:
https://stackoverflow.com/questions/3898665/what-is-in-bash

File: tests/e2e/containers/containers.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply tests/e2e/containers/containers.go:103]: I tried the suggestion and it seems that we indeed expect correct data to be returned via stderr sometimes

**Comment by p0mvn** [comment]
Thanks for the reviews, all comments are addressed so I'm going to merge this.
